modules/packages/Sort.chpl

```diff
@@ -802,6 +870,377 @@ proc selectionSort(Data: [?Dom] ?eltType, comparator:?rec=defaultComparator)
     compilerError("selectionSort() requires 1-D array");
 }
 
+pragma "no doc"
+proc shellSort(Data: [?Dom] ?eltType, comparator:?rec=defaultComparator,
+               start=Dom.low, end=Dom.high)
+  where !Dom.stridable
+{
+  chpl_check_comparator(comparator, eltType);
+
+  // Based on Sedgewick's Shell Sort -- see
+  // Analysis of Shellsort and Related Algorithms 1996
+  // and see Marcin Ciura - Best Increments for the Average Case of Shellsort
+  // for the choice of these increments.
+  var start_n = start;
+  var end_n = end;
+  var n = 1 + end_n - start_n;
+  var js,h,hs:int;
+  var v,tmp:Data.eltType;
+  const incs = (701, 301, 132, 57, 23, 10, 4, 1);
+  const nincs = incs.size;
+  for k in 1..nincs {
+    h = incs[k];
+    hs = h + start_n;
+    for is in hs..end_n {
+      v = Data[is];
+      js = is;
+      while js >= hs && chpl_compare(v,Data[js-h],comparator) < 0 {
+        Data[js] = Data[js - h];
+        js -= h;
+      }
+      Data[js] = v;
+    }
+  }
+}
+
+pragma "no doc"
+/* Error message for multi-dimension arrays */
+proc shellSort(Data: [?Dom] ?eltType, comparator:?rec=defaultComparator)
+  where Dom.rank != 1 {
+    compilerError("shellSort() requires 1-D array");
+}
+
+
+// This one would be harder to change since the code assumes
+// that 8 is a multiple of it, and 4 probably performs poorly.
+private param RADIX_BITS = 8;
+
+pragma "no doc"
+record MSBRadixSortSettings {
+  // TODO: move these as arguments to the radix sort fn?
+  param DISTRIBUTE_BUFFER = 5;
+  const sortSwitch = 256;
+  const minForTask = 256;
+  param CHECK_SORTS = false;
+  param progress = false;
+  const alwaysSerial = false;
+  const maxTasks = here.numPUs(logical=true);
+}
+
+// startbit is starting from 0
+// bin 0 is for the end was reached
+// bins 1..256 are for data with next part 0..255.
+//
+// returns (bin, ubits)
+pragma "no doc"
+inline
+proc binForRecordKeyPart(a, criterion, startbit:int)
+{
+  // We have keyPart(element, start):(section:int(8), part:int/uint)
+  var testRet: criterion.keyPart(a, 1).type;
```

Comment:
If I read correctly, all the variables in this function can be `const` except for `ubits`.